File: sdks/cpp/src/sak_mcp_client.cpp
#include "swissarmynoife/sak_mcp_client.hpp"

#include <httplib.h>
#include <stdexcept>

namespace swissarmynoife {
namespace {

void parse_mcp_url(const std::string& base, std::string& host, int& port, std::string& path) {
  auto u = base;
  while (!u.empty() && u.back() == '/') u.pop_back();
  std::string rest = u;
  if (rest.rfind("https://", 0) == 0) rest = rest.substr(8);
  else if (rest.rfind("http://", 0) == 0) rest = rest.substr(7);
  auto slash = rest.find('/');
  auto authority = slash == std::string::npos ? rest : rest.substr(0, slash);
  path = slash == std::string::npos ? "/" : rest.substr(slash);
  auto colon = authority.rfind(':');
  if (colon != std::string::npos) {
    host = authority.substr(0, colon);
    port = std::stoi(authority.substr(colon + 1));
  } else {
    host = authority;
    port = 80;
  }
}

}  // namespace

SakMcpClient::SakMcpClient(std::string base_url) {
  if (base_url.empty()) base_url = "http://127.0.0.1:8080/mcp";
  parse_mcp_url(base_url, host_, port_, path_);
}

nlohmann::json SakMcpClient::initialize() {
  auto result = rpc("initialize", {
                                    {"protocolVersion", "2024-11-05"},
                                    {"capabilities", nlohmann::json::object()},
                                    {"clientInfo", {{"name", "swissarmynoife-cpp"}, {"version", "0.1.0"}}},
                                });
  httplib::Client cli(host_, port_);
  httplib::Headers headers{{"Content-Type", "application/json"},
                           {"Accept", "application/json, text/event-stream"}};
  if (!token_.empty()) headers.emplace("Authorization", "Bearer " + token_);
  if (session_id_) headers.emplace("mcp-session-id", *session_id_);
  cli.Post(path_, headers, nlohmann::json({{"jsonrpc", "2.0"}, {"method", "notifications/initialized"}}).dump(),
           "application/json");
  initialized_ = true;
  return result;
}

std::string SakMcpClient::ping() { return extract_ping_text(tools_call("ping")); }

nlohmann::json SakMcpClient::tools_list() {
  ensure_session();
  return rpc("tools/list");
}

nlohmann::json SakMcpClient::catalog_list() { return tools_call("catalog_list"); }

void SakMcpClient::ensure_session() {
  if (!auto_initialize_ || initialized_) return;
  initialize();
}

nlohmann::json SakMcpClient::tools_call(const std::string& name, const nlohmann::json& arguments) {
  ensure_session();
  return rpc("tools/call", {{"name", name}, {"arguments", arguments}});
}

nlohmann::json SakMcpClient::rpc(const std::string& method, const nlohmann::json& params) {
  rpc_id_++;
  nlohmann::json payload = {{"jsonrpc", "2.0"}, {"id", rpc_id_}, {"method", method}, {"params", params}};
  httplib::Client cli(host_, port_);
  httplib::Headers headers{{"Content-Type", "application/json"},
                           {"Accept", "application/json, text/event-stream"}};
  if (!token_.empty()) headers.emplace("Authorization", "Bearer " + token_);
  if (session_id_) headers.emplace("mcp-session-id", *session_id_);
  auto res = cli.Post(path_, headers, payload.dump(), "application/json");
  if (!res || res->status < 200 || res->status >= 300) {
    throw std::runtime_error(std::to_string(res ? res->status : 0) + ": " + (res ? res->body : "no response"));
  }
  auto body = nlohmann::json::parse(res->body);
  if (!session_id_) {
    auto it = res->headers.find("mcp-session-id");
    if (it != res->headers.end() && !it->second.empty()) session_id_ = it->second;
  }
  if (!session_id_) session_id_ = session_id_from_body(body);
  if (body.contains("error")) {
    auto err = body["error"];
    auto msg = err.is_object() && err.contains("message") ? err["message"].get<std::string>() : err.dump();
    throw std::runtime_error("MCP " + method + " failed: " + msg);
  }
  return body.contains("result") ? body["result"] : body;
}

std::optional<std::string> SakMcpClient::session_id_from_body(const nlohmann::json& body) {
  for (auto key : {"sessionId", "session_id", "mcp-session-id"}) {
    if (body.contains(key) && body[key].is_string() && !body[key].get<std::string>().empty()) {
      return body[key].get<std::string>();
    }
  }
  if (body.contains("result") && body["result"].is_object()) {
    auto result = body["result"];
    for (auto key : {"sessionId", "session_id", "mcp-session-id"}) {
      if (result.contains(key) && result[key].is_string() && !result[key].get<std::string>().empty()) {
        return result[key].get<std::string>();
      }
    }
  }
  return std::nullopt;
}

std::string SakMcpClient::extract_ping_text(const nlohmann::json& result) {
  if (result.is_string()) return result.get<std::string>();
  if (result.is_object() && result.contains("content") && result["content"].is_array()) {
    for (auto& item : result["content"]) {
      if (item.is_object() && item.contains("text") && item["text"].is_string()) {
        return item["text"].get<std::string>();
      }
    }
  }
  return result.dump();
}

}  // namespace swissarmynoife

File: sdks/php/src/SakMcpClient.php
<?php

declare(strict_types=1);

namespace SwissArmyNoife\Sdk;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use Psr\Http\Message\ResponseInterface;

/** Streamable HTTP MCP client (sak335-c). */
final class SakMcpClient
{
    private const DEFAULT_MCP = 'http://127.0.0.1:8080/mcp';
    private const PROTOCOL = '2024-11-05';
    private const SESSION_HEADER = 'mcp-session-id';
    private const ACCEPT = 'application/json, text/event-stream';

    private string $baseUrl;
    private Client $http;
    private ?string $token = null;
    private bool $autoInitialize = true;
    private int $rpcId = 0;
    private ?string $sessionId = null;
    private bool $initialized = false;

    public function __construct(?string $baseUrl = null, ?Client $http = null)
    {
        $u = ($baseUrl === null || trim($baseUrl) === '') ? self::DEFAULT_MCP : $baseUrl;
        $this->baseUrl = rtrim($u, '/');
        $this->http = $http ?? new Client(['timeout' => 30.0, 'http_errors' => false]);
    }

    public function setToken(?string $token): void
    {
        $this->token = $token;
    }

    public function setAutoInitialize(bool $autoInitialize): void
    {
        $this->autoInitialize = $autoInitialize;
    }

    public function getSessionId(): ?string
    {
        return $this->sessionId;
    }

    /** @return mixed */
    public function initialize(): mixed
    {
        $result = $this->rpc('initialize', [
            'protocolVersion' => self::PROTOCOL,
            'capabilities' => new \stdClass(),
            'clientInfo' => [
                'name' => 'swissarmynoife-php',
                'version' => SdkInfo::VERSION,
            ],
        ]);
        $this->post(['jsonrpc' => '2.0', 'method' => 'notifications/initialized'], true);
        $this->initialized = true;

        return $result;
    }

    public function ping(): string
    {
        return $this->extractPingText($this->toolsCall('ping'));
    }

    /** @return mixed */
    public function toolsList(): mixed
    {
        $this->ensureSession();

        return $this->rpc('tools/list');
    }

    /** @return mixed */
    public function catalogList(): mixed
    {
        return $this->toolsCall('catalog_list');
    }

    private function ensureSession(): void
    {
        if (!$this->autoInitialize || $this->initialized) {
            return;
        }
        $this->initialize();
    }

    /**
     * @param array<string, mixed> $arguments
     * @return mixed
     */
    private function toolsCall(string $name, array $arguments = []): mixed
    {
        $this->ensureSession();

        return $this->rpc('tools/call', ['name' => $name, 'arguments' => $arguments]);
    }

    /**
     * @param array<string, mixed> $params
     * @return mixed
     */
    private function rpc(string $method, array $params = []): mixed
    {
        $this->rpcId++;
        $payload = [
            'jsonrpc' => '2.0',
            'id' => $this->rpcId,
            'method' => $method,
            'params' => $params,
        ];
        $res = $this->post($payload, false);
        /** @var array<string, mixed> $body */
        $body = json_decode((string) $res->getBody(), true, 512, JSON_THROW_ON_ERROR);
        $this->captureSession($res, $body);
        if (isset($body['error'])) {
            $err = $body['error'];
            $msg = is_array($err) ? ($err['message'] ?? json_encode($err)) : (string) $err;
            throw new \RuntimeException("MCP {$method} failed: {$msg}");
        }

        return $body['result'] ?? $body;
    }

    /**
     * @param array<string, mixed> $payload
     * @throws GuzzleException
     */
    private function post(array $payload, bool $notification): ResponseInterface
    {
        $headers = [
            'Content-Type' => 'application/json',
            'Accept' => self::ACCEPT,
        ];
        if ($this->token !== null && $this->token !== '') {
            $headers['Authorization'] = 'Bearer ' . $this->token;
        }
        if ($this->sessionId !== null && $this->sessionId !== '') {
            $headers[self::SESSION_HEADER] = $this->sessionId;
        }
        $res = $this->http->post($this->baseUrl, [
            'headers' => $headers,
            'json' => $payload,
            'http_errors' => false,
        ]);
        $code = $res->getStatusCode();
        if ($notification && ($code === 200 || $code === 202)) {
            return $res;
        }
        if ($code < 200 || $code >= 300) {
            throw new \RuntimeException($code . ': ' . (string) $res->getBody());
        }

        return $res;
    }

    /** @param array<string, mixed> $body */
    private function captureSession(ResponseInterface $res, array $body): void
    {
        if ($this->sessionId === null || $this->sessionId === '') {
            $sid = $res->getHeaderLine(self::SESSION_HEADER);
            if (trim($sid) !== '') {
                $this->sessionId = trim($sid);
            }
        }
        if ($this->sessionId === null || $this->sessionId === '') {
            $fromBody = $this->sessionIdFromBody($body);
            if ($fromBody !== null) {
                $this->sessionId = $fromBody;
            }
        }
    }

    /** @param array<string, mixed> $body */
    private function sessionIdFromBody(array $body): ?string
    {
        foreach (['sessionId', 'session_id', 'mcp-session-id'] as $key) {
            if (isset($body[$key]) && is_string($body[$key]) && trim($body[$key]) !== '') {
                return trim($body[$key]);
            }
        }
        $result = $body['result'] ?? null;
        if (is_array($result)) {
            foreach (['sessionId', 'session_id', 'mcp-session-id'] as $key) {
                if (isset($result[$key]) && is_string($result[$key]) && trim($result[$key]) !== '') {
                    return trim($result[$key]);
                }
            }
        }

        return null;
    }

    private function extractPingText(mixed $result): string
    {
        if (is_string($result)) {
            return $result;
        }
        if (!is_array($result)) {
            return (string) json_encode($result);
        }
        $content = $result['content'] ?? null;
        if (is_array($content)) {
            foreach ($content as $item) {
                if (is_array($item) && isset($item['text']) && is_string($item['text'])) {
                    return $item['text'];
                }
            }
        }

        return (string) json_encode($result);
    }
}
